influxql/iterator.go

```diff
@@ -47,6 +47,82 @@ func (a Iterators) filterNonNil() []Iterator {
 	return other
 }
 
+// castIterators casts an array of iterators to a single type.
+// An iterator type is chosen using this hierarchy:
+//   float > integer > string > boolean
+// Iterators that are not compatible or cannot be cast to the
+// chosen iterator type are closed and dropped.
+func castIterators(inputs []Iterator) interface{} {
+	var types struct {
+		floats   int
+		integers int
+		strings  int
+		booleans int
+	}
+
+	for _, input := range inputs {
+		switch input.(type) {
+		case FloatIterator:
+			types.floats++
+		case IntegerIterator:
+			types.integers++
+		case StringIterator:
+			types.strings++
+		case BooleanIterator:
+			types.booleans++
+		}
+	}
+
+	if types.floats > 0 {
+		itrs := make([]FloatIterator, 0, types.floats+types.integers)
+		for _, input := range inputs {
+			switch input := input.(type) {
+			case FloatIterator:
+				itrs = append(itrs, input)
+			case IntegerIterator:
+				itrs = append(itrs, &castFloatIterator{input: input})
+			default:
+				input.Close()
+			}
+		}
+		return itrs
+	} else if types.integers > 0 {
+		itrs := make([]IntegerIterator, 0, types.integers)
+		for _, input := range inputs {
+			switch input := input.(type) {
+			case IntegerIterator:
+				itrs = append(itrs, input)
+			default:
+				input.Close()
+			}
+		}
+		return itrs
+	} else if types.strings > 0 {
+		itrs := make([]StringIterator, 0, types.strings)
+		for _, input := range inputs {
+			switch input := input.(type) {
+			case StringIterator:
+				itrs = append(itrs, input)
+			default:
+				input.Close()
+			}
+		}
+		return itrs
+	} else if types.booleans > 0 {
+		itrs := make([]BooleanIterator, 0, types.booleans)
+		for _, input := range inputs {
+			switch input := input.(type) {
+			case BooleanIterator:
+				itrs = append(itrs, input)
+			default:
+				input.Close()
+			}
+		}
+		return itrs
+	}
+	return inputs
+}
```

Comment:
I'm not sure I see any meaningful difference between the two. I think the optimization in terms of keeping memory usage smaller is worth the function being a bit longer personally.

This was the first solution I was considering, but I figured the extra optimization was worth it and the extra functions required to split the functionality wasn't worth it.